File: Strings/CPP/ZigZagConversion.cpp
/*
LeetCode Question:
https://leetcode.com/problems/zigzag-conversion/
*/
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class ZigZagConversion {
public:
    string convert(const string& s, int numRows) {
        if (numRows <= 1 || s.length() <= numRows) {
            return s;
        }

        vector<string> rows(numRows);
        int currentRow = 0;
        bool goingDown = false;

        for (char c : s) {
            rows[currentRow] += c;
            if (currentRow == 0 || currentRow == numRows - 1) {
                goingDown = !goingDown;
            }
            currentRow += goingDown ? 1 : -1;
        }

        string result;
        for (const string& row : rows) {
            result += row;
        }

        return result;
    }
};

int main() {
    ZigZagConversion solution;
    string s = "PAYPALISHIRING";
    int numRows = 3;
    cout << solution.convert(s, numRows) << endl; 
    return 0;
}

File: Strings/Java/ZigZagConversion.java
/*
LeetCode Question:
https://leetcode.com/problems/zigzag-conversion/
*/
import java.util.ArrayList;
import java.util.List;

public class ZigZagConversion {
    public String convert(String s, int numRows) {
        if (numRows <= 1 || s.length() <= numRows) {
            return s;
        }

        List<StringBuilder> rows = new ArrayList<>();
        for (int i = 0; i < numRows; i++) {
            rows.add(new StringBuilder());
        }

        int currentRow = 0;
        boolean goingDown = false;

        for (char c : s.toCharArray()) {
            rows.get(currentRow).append(c);
            if (currentRow == 0 || currentRow == numRows - 1) {
                goingDown = !goingDown;
            }
            currentRow = currentRow + goingDown ? 1 : -1;
        }

        StringBuilder result = new StringBuilder();
        for (StringBuilder row : rows) {
            result.append(row);
        }

        return result.toString();
    }

    public static void main(String[] args) {
        ZigZagConversion solution = new ZigZagConversion();
        String s = "PAYPALISHIRING";
        int numRows = 3;
        System.out.println(solution.convert(s, numRows));
    }
}
